### PR TITLE
Fix #38 Message alignment

### LINKED ISSUE #38: [Message:005]: Send and receive event alignment constraint
When a message has its send, and receive events aligned (same vertical position), this alignment will be always maintained, except for explicitly move or reorder the message end along the lifeline. Under any other circumstance the aligned will be maintained.

### PR DESCRIPTION
**Pull request by cdamus**
Implement constraint of message alignment when reconnecting an end, according to message sort:

- synchronous messages must be horizontal
- asynchronous messages cannot slope up

### CODE REVIEW

**Comment by cdamus** [comment]
Agreed on moving by grabbing the message somewhere along the line segment between its ends.

Also, as this issue really is (in my interpretation) about the visual orientation of the message, the editing behaviour that it supports should not have any impact on the semantics.  So, in particular, it must not provide for semantic re-ordering of message ends with respect to other interaction fragments.  So, I need to continue this PR branch with two further changes:
* restrict the movement of a message end such that it does not leap-frog any other interaction fragment on the lifeline
* if a message end is moved into proximity with another interaction fragment on the lifeline, ensure a sufficient padding by nudging the layout (this implies a buffer on how close an end can be brought to the predecessor fragment on the lifeline, because we mustn't nudge that fragment up; it makes sense only to nudge the successors of a message end

**Note** that none of this supports interaction of message ends with execution specifications:  message ends here are only attaching to the lifeline body, and we are not dealing with executions starting or finishing with a message end.

**Comment by cdamus** [comment]
Okay, I have fixed anchoring of messages on execution specifications (especially the asynchronous ones) in commit 3b93ffd14619e1ab31eba5dd8080911583e8f8d3 and merged changes from master to resolve conflicts.  JUnit tests included.

Could I get another review from @planger or @rschnekenbu ? It would be nice to be able to complete this PR before too many more changes need conflict resolution.  😉  Thanks.

**Comment by cdamus** [comment]
Trucking along, the latest commit adds support discussed earlier for moving messages up and down (without changing their shape) simply by grabbing them with the mouse pointer and dragging them.

**Comment by planger** [comment]
I created a new task for the first bullet: #323
The second bullet is actually already captured: #48

**Comment by planger** [comment]
Unless @rschnekenbu has more feedback, I'm fine with merging this. Thanks!

File: plugins/org.eclipse.papyrus.uml.diagram.sequence.runtime/src/org/eclipse/papyrus/uml/diagram/sequence/runtime/internal/edit/policies/AbstractSequenceGraphicalNodeEditPolicy.java [hunk 1]
```diff
@@ -245,6 +246,11 @@ protected FeedbackHelper getFeedbackHelper(CreateConnectionRequest request) {
 	protected Command getReconnectSourceCommand(ReconnectRequest request) {
 		Command result = super.getReconnectSourceCommand(request);
 
+		if (isForce(request)) {
+			// Just reconnect this end; no sloping constraints required
+			return result;
```
Comment (rschnekenbu) [inline plugins/org.eclipse.papyrus.uml.diagram.sequence.runtime/src/org/eclipse/papyrus/uml/diagram/sequence/runtime/internal/edit/policies/AbstractSequenceGraphicalNodeEditPolicy.java]: When grabbing a message from the center, the request is forced. Thus we do not check validity of the move using _constrainReconnection_. In this case, we can grab and reorder graphically messages, while their ends are semantically not reordered. Is this done on purpose?

Comment (cdamus) [reply plugins/org.eclipse.papyrus.uml.diagram.sequence.runtime/src/org/eclipse/papyrus/uml/diagram/sequence/runtime/internal/edit/policies/AbstractSequenceGraphicalNodeEditPolicy.java]: The constraints referenced here are the constraints on how a message may be sloped:  not at all for synchronous messages and forwards in time only for asynchronous messages.

The re-ordering constraint is implemented in the calculation of the semantic reconnection command in the graphical node edit-policy.  So, the feedback should show the connection moving but then the mouse pointer may show the no-smoking cursor.

Comment (rschnekenbu) [reply plugins/org.eclipse.papyrus.uml.diagram.sequence.runtime/src/org/eclipse/papyrus/uml/diagram/sequence/runtime/internal/edit/policies/AbstractSequenceGraphicalNodeEditPolicy.java]: I do not get this no-smoking cursor. I can currently shift a sloped message, so the end of this message is graphically placed after the following message for example.  Debugging lead me to that place, I was expecting to go though method constrainReconnection(ReconnectRequest request, MMessageEnd end) defined in commit #7003616, but this is short circuited by the force parameter on the request

Comment (cdamus) [reply plugins/org.eclipse.papyrus.uml.diagram.sequence.runtime/src/org/eclipse/papyrus/uml/diagram/sequence/runtime/internal/edit/policies/AbstractSequenceGraphicalNodeEditPolicy.java]: Ah, right.  Sorry, I thought we were looking at the feed-back helper here.  I'll look into it.

**Pull request merged**